### DISCUSSION TITLE
Force watch state from backend to db

### DISCUSSION THREAD

**Discussion by Nyxtorm** [arabcoders/watchstate#428] [General] [answered]
Hello,

I discovered watchstate which is great but I have a problem.

I currently have two backends:

- Plex, I want it to always be master
- Emby, I want it to always retrieve information from Plex, even if changes are made.

To do this :

- I import information from Plex:

```
docker exec -ti watchstate console state:import -vfs plex
```

- Then I import Emby metadata only:

```
docker exec -ti watchstate console state:import -vfs emby --metadata-only
```

Then I force a full export to Emby:

```
docker exec -ti watchstate console state:export -vifs emby
```

My problem is that if I mark an element as unplayed on Plex, from the moment the watchstate database has seen it as played, I systematically get this error:

```
WARNING: MAPPER: Watch state conflict detected in [plex] [XXXXXXX (YYYY)] [unplayed] vs db [played].
```

**How can I force the "db" to integrate the new plex's unplayed status?**

I've tried searching in the FAQ or the built-in help without finding a solution.

Thanks in advance for your help!

**Comment by arabcoders** [comment]
Hello,

This should be easily supported as you discovered, The error you are receiving is most likely due to the plex timezone being different than that of the watchstate container, try matching them using the `WS_TZ` environment variable, and start fresh.. please report back if that fixes your problem.

**Comment by Nyxtorm** [comment]
Hi @arabcoders, 

Thank you for your prompt reply.

- The environment variable on watchstate :

```
$ echo $WS_TZ
Europe/Paris
```

- The Debian timezone where Plex is hosted:

```
# cat /etc/timezone
Europe/Paris
```

The only method I can think of, but it's still a "do-it-yourself" job, would be to filter out all the `MAPPER: Watch state conflict detected in` in the import's output and extract the element titles, then apply to each :

```
docker exec -ti watchstate console db:list --title="XXXXXXX" --mark-as unplayed
```

But even with the `-n` option, this requires interaction.

EDIT : The `-i` option in the export should ignore the dates, and therefore the timezone, no?

**Reply by arabcoders** [reply]
I would like to discover why the state conflict is being detected if only plex is updating the play state. I regularly mark stuff as unplayed and it's working as expected, are you by chance not using webhooks?

**Reply by Nyxtorm** [reply]
If it can helps, here is my configuration for Plex and Emby obfuscated : 

```
# docker exec -ti watchstate console config:view
┌───────────┬────────────────────────────────────────────────┐
│ Backend   │ Data (Filter: None)                            │
├───────────┼────────────────────────────────────────────────┤
│ plex_odin │ name: plex_odin                                │
│           │ type: plex                                     │
│           │ token: '*HIDDEN*'                              │
│           │ url: 'https://xxx.xxx.xxx:xxx'                 │
│           │ uuid: xxxxxx                                   │
│           │ user: xxxxxx                                   │
│           │ options:                                       │
│           │   plex_user_uuid: xxxxxx                       │
│           │   ADMIN_TOKEN: '*HIDDEN*'                      │
│           │ import:                                        │
│           │   enabled: true                                │
│           │   lastSync: '2024-04-29T15:32:08+02:00'        │
│           │ export:                                        │
│           │   enabled: true                                │
│           │   lastSync: '2024-04-29T15:06:06+02:00'        │
│           │ webhook:                                       │
│           │   match:                                       │
│           │     user: true                                 │
│           │     uuid: true                                 │
│           │   token: '*HIDDEN*'                            │
├───────────┼────────────────────────────────────────────────┤
│ emby_odin │ name: emby_odin                                │
│           │ type: emby                                     │
│           │ url: 'https://xxx.xxx.xxx:xxx'                 │
│           │ token: '*HIDDEN*'                              │
│           │ uuid: xxxxxx                                   │
│           │ user: xxxxxx                                   │
│           │ import:                                        │
│           │   enabled: false                               │
│           │   lastSync: '2024-04-29T15:11:41+02:00'        │
│           │ export:                                        │
│           │   enabled: true                                │
│           │   lastSync: '2024-04-29T15:06:06+02:00'        │
│           │ options:                                       │
│           │   IMPORT_METADATA_ONLY: true                   │
│           │ webhook:                                       │
│           │   match:                                       │
│           │     user: true                                 │
│           │     uuid: true                                 │
│           │   token: '*HIDDEN*'                            │
└───────────┴────────────────────────────────────────────────┘
```

**Reply by arabcoders** [reply]
Seems ok for me, is there any errors in `system:report`?

**Reply by Nyxtorm** [reply]
There seems to be no mistake.

Here is the report, which will provide more information on versions, etc...

```
[ Basic Report ]

WatchState Version: 20240428-7db51ea
PHP Version: 8.3.6
Timezone: Europe/Paris
Running in Container? Yes
Data Path: /config
Temp Path: /config
Database Migrated?: Yes
Has .env file? Yes
Is Tasks Runner working? Yes
Report Generated At: 2024-04-29T14:10:28+00:00

[ Backends ]

[ Plex (1.40.2.8395-c67dce28e) ==> plex_odin ]

Is backend URL HTTPS? Yes
Has Unique Identifier? Yes
Has User? Yes
Export Enabled? Yes
Time since last export? 2024-04-29T13:06:06+00:00
Play state import enabled? Yes
Metadata only import enabled? No
Time since last import? 2024-04-29T13:32:08+00:00
Has webhook token? Yes
Is webhook match user id enabled? Yes
Is webhook match backend unique id enabled? Yes
Has custom options? Yes
{"plex_user_uuid":"a012fd00e80e3f8b","ADMIN_TOKEN":"**HIDDEN**"}

[ Emby (4.9.0.16) ==> emby_odin ]

Is backend URL HTTPS? Yes
Has Unique Identifier? Yes
Has User? Yes
Export Enabled? Yes
Time since last export? 2024-04-29T13:06:06+00:00
Play state import enabled? Yes
Metadata only import enabled? Yes
Time since last import? 2024-04-29T13:11:41+00:00
Has webhook token? Yes
Is webhook match user id enabled? Yes
Is webhook match backend unique id enabled? Yes
Has custom options? Yes
{"IMPORT_METADATA_ONLY":true}

[ Tasks ]

[ Import ]

Is Task enabled? No

[ Export ]

Is Task enabled? No

[ Push ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? */10 * * * *
When is the next scheduled run? 2024-04-29T14:20:00+00:00

[ Progress ]

Is Task enabled? No

[ Backup ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 6 */3 * *
When is the next scheduled run? 2024-05-01T04:00:00+00:00

[ Prune ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 */12 * * *
When is the next scheduled run? 2024-04-29T22:00:00+00:00

[ Indexes ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 3 * * 3
When is the next scheduled run? 2024-05-01T01:00:00+00:00

[ Requests ]

Is Task enabled? Yes
Which flags are used to run the task? -v --no-stats
When the task scheduled to run at? */2 * * * *
When is the next scheduled run? 2024-04-29T14:12:00+00:00

[ Logs ]

[ /config/logs/app.20240429.log ]

app log file is empty or does not exists.

[ /config/logs/access.20240429.log ]

access log file is empty or does not exists.

[ /config/logs/task.20240429.log ]

--------------------------
Task: push (Started: Mon, 15:50:07 CEST)
Command: /opt/app/bin/console state:push -v
Exit Code: 0 (Ended: Mon, 15:50:08 CEST)
--------------------------
[2024-04-29T15:50:08+02:00] NOTICE: SYSTEM: No play state changes detected.

[ /config/logs/app.20240428.log ]

app log file is empty or does not exists.

[ /config/logs/access.20240428.log ]

access log file is empty or does not exists.

[ /config/logs/task.20240428.log ]

task log file is empty or does not exists.

<!-- Notice

Beware, while we try to make sure no sensitive information is leaked, it's possible
that some private information might be leaked via the logs section.
Please review the report before posting it.
-->
```

**Reply by arabcoders** [reply]
Thanks. for the report, If you want could you try the following, remove the watchstate.db, and do `state:import -f -s plex_odin`, after that try marking someone unwatched, and wait for the webhook to kick in there should be a task.YYMMDD.log which info when that happens.

**Reply by Nyxtorm** [reply]
What I did:

- Delete db/watchstate_v01.db
- Import only Plex in local db : `state:import -vfs plex_odin`
- Mark movie as unplayed on Plex
- Wait for push task (if you meant that with "_wait for the webhook to kick in_") , no output in task.xxxxxx log

Then, I checked more verbose manual push : 

```
# docker exec -ti watchstate console state:push -vvv --trace --debug
[2024-04-29T16:40:54+02:00] INFO: No items in the queue.
```

**Reply by arabcoders** [reply]
I've also pushed new change to allow changing the state without interaction using the `-n` flag as workaround while we diagnose the issue.

**Reply by arabcoders** [reply]
try `system:logs -t access` do you see any logs?

**Reply by Nyxtorm** [reply]
Thank you for your last action, I'll check this.

For logs :

```
# docker exec -ti watchstate console system:logs -t access
Log file [config/logs/access.20240429.log] does not exist or is empty. This means it has been pruned, the date is incorrect or nothing has written to that file yet.
```

Actually :

```
# docker exec -ti watchstate console system:logs --list
┌──────┬──────────┬─────────┬──────────────────────────┐
│ Type │ Tag      │ Size    │ Modified                 │
├──────┼──────────┼─────────┼──────────────────────────┤
│ task │ 20240429 │ 259.00B │ 2024-04-29 15:50:08 CEST │ => only old content before tests
└──────┴──────────┴─────────┴──────────────────────────┘
```

**Reply by arabcoders** [reply]
This probably means the import task is yet to run, i forgot plex doesn't send webhook events for marked as played/unplayed. you can manually trigger the task by `system:tasks --run --task import --save-log`

**Reply by Nyxtorm** [reply]
Actually, to reproduce easily, I just do:

- Delete the local database file
- New initial import with my test movie "unplayed" on Plex (only Plex backend): `state:import -vfs plex_odin`
- Mark movie as played, and renew same import. watchstate can see the change:

```
[2024-04-29T17:20:19+02:00] NOTICE: MAPPER: [plex_odin] Updated and marked [XXXXX] as [played].
```

- Just mark movie as unplayed on Plex and do the same import again:

```
[2024-04-29T17:20:32+02:00] WARNING: MAPPER: Watch state conflict detected in [plex_odin] [XXXXX] [unplayed] vs db [played].
```

It seems the conflict is between Plex and local file only. It could be great to force local database state in this case, especially with `-f` option.

**Reply by arabcoders** [reply]
Thanks for the reproduce, i have to leave for now, but i'll update you later, for now please update to the `arabcoders/watchstate:dev` to use the mark-as flag without interaction until i can trace why this issue is happening.

**Reply by Nyxtorm** [reply]
You're welcome, thank you. I'm already on the dev tag and up to date. 

In the meantime, I redefine local unplayed from Plex as follows before "real" import/export :

```bash
docker exec -ti watchstate console state:import -vfs plex_odin --dry-run | grep "MAPPER: Watch state conflict detected in" | awk -F '[][]' '{
    split($6, parts, "[()]");
    title=parts[1];
    year=parts[2];
    gsub(/[[:space:]]+$/, "", title);
    print "docker exec -t watchstate console db:list --title=\"" title "\" --year " year " --mark-as unplayed -n --exact"
}' | bash >> /dev/null
```

I use the year to filter more finely, as `--exact` does not seem to work here. For example `--title="Ghost"` also matches "Ghost in the Shell".

**Reply by arabcoders** [reply]
I probably could add the db item id for easier filtering.

**Reply by arabcoders** [reply, marked as answer]
@Nyxtorm I think i traced your issue, the code path you are hitting shouldn't normally be triggered in your case, it's happening due to using `-f` flag, enable the import task and let it run normally, and you should have mark as unplayed working as intended. In anycase i have added db item id in the log message, `vs db [(id):...` you can then `db:list --id (id) --mark-as unplayed -n` if you still want to use the `-f` flag.

I've also pushed new code to bypass the check if `-f` is present.

**Reply by Nyxtorm** [reply]
@arabcoders Thank you very much for these latest changes. I was able to achieve what I wanted for my Plex and Emby servers.
